### ISSUE TITLE
Make tester's package name detection work packages installed in a non standard location (migrated from Trac #1168)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#2721]
Original ticket http://projects.scipy.org/numpy/ticket/1168
Reported 2009-07-11 by atmention:GaelVaroquaux, assigned to unknown.

Right now, numpy's tester package name detection (used to do coverage analysis) fails for a package that is installed in a non-standard location. This is due to the following logics:
    def get_package_name(filepath):
        # find the package name given a path name that's part of the package
        fullpath = filepath[:]
        pkg_name = []
        while 'site-packages' in filepath:
            filepath, p2 = os.path.split(filepath)
            if p2 == 'site-packages':
                break
            pkg_name.append(p2)
which fails if the package is installed outside a 'site-package' directory.

I attach a patch that fixes this problem (works for me), and I believe is more robust.

**Comment by thouis** [comment]
Attachment in Trac by atmention:GaelVaroquaux, 2009-07-11: [tmp.diff](http://projects.scipy.org/numpy/attachment/ticket/1168/tmp.diff)

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2009-07-12

Looks good to me. Tested. Committed in r7132

**Comment by thouis** [comment]
Comment in Trac by atmention:huard, 2009-08-12

The applied patch removes support for packages passed a a string. Here is a patch that fixes the issue I'm seeing.

```
Index: numpy/testing/nosetester.py
===================================================================
--- numpy/testing/nosetester.py (revision 7301)
+++ numpy/testing/nosetester.py (working copy)
atmention:atmention: -125,6 +125,9 atmention:atmention:
         elif isinstance(package, type(os)):
             package_path = os.path.dirname(package.__file__)
             package_name = getattr(package, '__name__', None)
+        else:
+            package_path = package
+
         self.package_path = package_path

         # find the package name under test; this name is used to limit coverage
```

**Comment by thouis** [comment]
Attachment in Trac by atmention:jseabold, 2009-08-23: [tmp_diff2.txt](http://projects.scipy.org/numpy/attachment/ticket/1168/tmp_diff2.txt)

**Comment by thouis** [comment]
Comment in Trac by atmention:jseabold, 2009-08-23

I also just wrote a patch for this.  This last patch for if packages is a string doesn't check if it's a string before the final else, which might be helpful.

I also made get_package_name more robust for files installed in dist-packages and if the package is installed in an unzipped egg (discovered while working on my scikit installation).  It works for me.

Also improved the docstring for the test method in NoseTester.

Patch should be attached if I did everything right.
